Feature/binary

blairlyons opened the pull request:
Problem
=======
SimulariumIO only outputs the JSON version of .simularium format, it would be nice to also output binary.
closes #30 

Solution
========
Split file writing functionality from `TrajectoryConverter` into a few writer objects, including abstract base `Writer`, `JSONWriter`, and `BinaryWriter`. Each of those implements `Writer.save()`, which replaces the old `TrajectoryConverter.write_external_JSON()`. The old `TrajectoryConverter.write_JSON()` functions are replaced with `TrajectoryConverter.save()` which uses the writers and defaults to writing binary (but can set `binary=False` to write JSON).

**Note** this changes the `TrajectoryConverter` API, renaming `write_JSON()` to `save()` and removing `write_external_JSON()`. There's also a new module `writers`.

**Note** Production simularium.allencell.org needs updates that are currently on staging before this PR can be merged

## Type of change
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update
- [x] This change requires updated or new tests

Change summary:
---------------
* move JSON writing to `Writer` and `JsonWriter`
* add `BinaryWriter` to write to binary format
* split binary files so they don't exceed size limit by bytes
* update `TrajectoryConverter` functions
* add tests that check the float values in the buffer before writing binary are correct
* update tests and docs
* update jupyter notebook examples
* add spec to the file format for binary
* update binary writing to match new spec

Keyfiles:
-----------------------
I tried to comment on the diff too to make this easier to review.

1. file_format.md - updated spec to include binary files
2. simulariumio/__init__.py - added `writers` to main API
3. simulariumio/constants.py - added constants for binary format
4. simulariumio/trajectory_converter.py - moved `_read_trajectory_data()` to `JsonWriter.format_trajectory_data()` and moved part of `_get_spatial_bundle_data_subpoints()` to `Writer._get_frame_buffer()` so it can also be used by `BinaryWriter`. Updated `to_JSON()` to use `JsonWriter`, changed `write_JSON()` to `save(binary=True)`, and removed `write_external_JSON()` since that's now handled by `JsonWriter.save()`.
5. simulariumio/writers/binary_chunk.py - added this object to store info about a chunk of data that will be saved to binary, which allows the trajectory to be split into multiple files if the data is too large
6. simulariumio/writers/binary_values.py - added this object to store values to be saved in binary and their corresponding format strings
7. simulariumio/writers/binary_writer.py - format data and save as binary in `save()`
8. simulariumio/writers/json_writer.py - moved `TrajectoryData` reading and JSON writing from `TrajectoryConverter` here, except moved parts of the old `TrajectoryConverter._read_trajectory_data()` to `Writer` in order to reuse it for writing binary.
9. simulariumio/writers/writer.py - abstract base `Writer` that defines an API with `format_trajectory_data()` and `save()` and also has a few functions used by both implementing classes.

blairlyons commented:
from meeting early September:
- trajectoryInfo and plotData blocks are still in JSON only, can be saved to same file with binary spatialData
- to pack different types into binary: https://docs.python.org/3/library/struct.html#struct.pack (from Dan)

blairlyons commented on file_format.md:
```diff
@@ -1,6 +1,10 @@
-# .simularium JSON Format
+# .simularium File Format
```
the file format is no longer just JSON, update descriptions of structure and files

blairlyons commented on file_format.md:
```diff
@@ -37,10 +41,11 @@ JSON files accepted by the simularium-viewer contain the following data in JSON
       * visualization type
         * 1000 = default, rendered with PDB or mesh
         * 1001 = fiber, rendered as a line
+      * agent instance ID - integer number ID for this agent instance
```
whoops this was missing

blairlyons commented on file_format.md:
```diff
@@ -37,10 +41,11 @@ JSON files accepted by the simularium-viewer contain the following data in JSON
       * position X, Y, Z - agent’s 3D position
       * rotation X, Y, Z - euler angles for agent’s orientation
-      * radius - the radius the agent occupies, used for drawing a sphere
+      * radius - the radius the agent occupies, used for drawing a sphere or for scaling other geometry representations
```
update for geometry

blairlyons commented on file_format.md:
````diff
@@ -235,3 +243,55 @@ JSON files accepted by the simularium-viewer contain the following data in JSON
     }
 }
 ```
+
````
added spec for binary files, which we discussed here https://github.com/simularium/simularium-viewer/issues/123

blairlyons commented on simulariumio/__init__.py:
```diff
@@ -30,3 +30,8 @@ def get_module_version():
     HistogramPlotData,
 )
 from .constants import DISPLAY_TYPE  # noqa: F401
+
```
make the writers accessible from the main module

blairlyons commented on simulariumio/constants.py:
```diff
@@ -74,6 +75,58 @@ class CURRENT_VERSION:
 
 DEFAULT_BOX_SIZE = 100.0 * np.ones(3)
 
+
```
added constants for binary format

blairlyons commented on simulariumio/trajectory_converter.py:
```diff
@@ -56,374 +53,6 @@ def __init__(self, input_data: TrajectoryData):
         """
         self._data = input_data
 
```
moved these functions to the writers, split up between JsonWriter and base Writer (for functions that are reused by BinaryWriter)

blairlyons commented on simulariumio/trajectory_converter.py:
```diff
@@ -469,9 +98,12 @@ def add_number_of_agents_plot(
-        agent_data: AgentData
-            The data shaped as an AgentData object
-            Default: None (use the currently loaded data)
+        plot_title: str
```
whoops this docstring was wrong

blairlyons commented on simulariumio/trajectory_converter.py:
```diff
@@ -507,40 +139,35 @@ def to_JSON(self):
 
         """
-        buffer_data = TrajectoryConverter._read_trajectory_data(self._data)
+        buffer_data = JsonWriter.format_trajectory_data(self._data)
```
moved this functionality to JsonWriter, I think this function could still be convenient

blairlyons commented on simulariumio/trajectory_converter.py:
```diff
@@ -507,40 +139,35 @@ def to_JSON(self):
         Return the current simularium data in JSON format
 
         """
```
updated these functions since they've been moved to the writers. `TrajectoryConverter.write_external_JSON()` is now `JsonWriter.save()`

blairlyons commented on simulariumio/writers/binary_chunk.py:
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env python
```
BinaryChunk is used to determine how many files to save and how to chunk up the data, if the data is larger than the max file size

toloudis replied:
The above is a perfect line for a comment in the code , right where it says "Object to store size info for a binary file"

blairlyons commented on simulariumio/writers/binary_values.py:
```diff
@@ -0,0 +1,35 @@
+#!/usr/bin/env python
```
`BinaryValues` stores values to be saved in binary and their corresponding format strings

blairlyons commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,417 @@
+        """
+        first_frame_offset = constant_first_frame_offset + chunk.n_frames
+        chunk.frame_offsets = [
+            offset + first_frame_offset for offset in chunk.frame_offsets
```
@toloudis should the frame offsets be the global offset from the start of the file, or the offset within the spatial data block? I assumed global would be easier for parsing, but wanted to check

toloudis replied:
If this is what I think it is,  then I think it should be offsets within the spatial data block.  Mainly so that we aren't changing too many things. We already have code that expects these offsets to be that.  I'll be able to answer more definitively when I really dig into this code in a couple days.

blairlyons replied:
ok let me know and I can make that change

toloudis replied:
yes please

toloudis commented:
A couple of issues as I start writing a typescript parser for this.  
1.  The example file you gave me is little endian.  Make sure that the python code is enforcing/guaranteeing this.  
2. For the headers of individual blocks, the spec says that blocktype comes before block size but in the file you gave me, block size came first.
3. The block size saved into the block actually is the block offset, not its size. (This one renders the file unreadable till it's fixed)
4. Make sure (I haven't checked) that the block size includes the block header data (its size and type). We should note that in the spec.  To get the raw data from the block you have to read (blocksize-blockheadersize) bytes from (blockoffset+blockheadersize) offset.

toloudis commented:
One more thing I have noticed.  I think for the spatial data, it will be convenient (and possibly necessary) to have a byte length as a part of each frame in the spatial data info.  I think the back end needs to be able to transmit bundles of data that don't necessarily end on frame boundaries, and having that byte length would allow us to know where the frame was split. (Much like the EOF signature).

blairlyons commented:
@toloudis ok 1 is done.

For 2-4, my tests are checking those things, or I just checked them, and they should be correct. But I didn't do the offsets correctly, I forgot to go back and multiply to put them in bytes, seems like that would do it. Now all offsets and lengths are in bytes.

I also updated the offsets within the spatial block to be local to the block and added frame lengths. Just like for the blocks in the file header they are intermixed with frame offsets (i.e. frame 0 offset, frame 0 length, frame 1 offset, frame 1 length, ...).

Data is updated at same URL. Let me know if anything else is off!

blairlyons commented on simulariumio/data_objects/agent_data.py:
```diff
@@ -429,12 +429,19 @@ def from_dimensions(
+        Use n_timesteps to limit times if it has been provided
+        """
+        return self.n_timesteps if self.n_timesteps >= 0 else len(self.times)
+
```
this is convenience for `BinaryWriter`

toloudis replied:
as a reader of this code, I think it would help if you had consistent naming: choose one of `timesteps`, `steps` or `frames` if possible.  All 3 are used in the above function and it's confusing.

blairlyons commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+            buffer_data = json.loads(input_file.get_contents())
+        except UnicodeDecodeError:
+            print("Reading Simularium binary -------------")
+            buffer_data = FileConverter._load_binary(input_file)
```
update to allow `FileConverter` to read in binary files

toloudis replied:
The above try/except would probably be a lot more efficient if you could just read the first 16 bytes of the file and compare them with "SIMULARIUMBINARY" with a regular old if statement.  That check could be encapsulated as a little function in FileConverter, like `if is_binary_file(input_file):`

toloudis replied:
is it `get_contents` that raises the UnicodeDecodeError ?

blairlyons replied:
added `InputFileData._is_binary_file()` which reads the header to check for binary

blairlyons commented on simulariumio/tests/conftest.py:
```diff
@@ -566,3 +566,203 @@ def mixed_agents() -> TrajectoryData:
+        assert False not in np.isclose(
+            np.array(test_frame["data"]), np.array(expected_frame["data"])
+        )
+    assert test_buffer["plotData"] == expected_buffer["plotData"]
```
this allows testing floats for isclose() rather than equality

toloudis commented on simulariumio/constants.py:
```diff
@@ -74,6 +75,57 @@ class CURRENT_VERSION:
+class BINARY_SETTINGS:
+    HEADER: str = "SIMULARIUMBINARY"
+    VERSION: int = 2
+    MAX_BYTES: int = 4000000000  # 4GB is max for one file
```
technically this is (4*1024*1024*1024)-1 but your number is more conservative.

toloudis commented on simulariumio/constants.py:
```diff
@@ -74,6 +75,57 @@ class CURRENT_VERSION:
+    BYTES_PER_VALUE: int = 4
+    BLOCK_OFFSET_BYTE_ALIGNMENT: int = 4
+
+    def HEADER_N_INT_VALUES() -> int:
```
can this be defined as an expression rather than a function?

toloudis commented on simulariumio/constants.py:
```diff
@@ -74,6 +75,57 @@ class CURRENT_VERSION:
+            + BINARY_SETTINGS.N_BLOCKS * BINARY_SETTINGS.HEADER_N_VALUES_PER_BLOCK
+        )
+
+    def DEFAULT_BLOCK_TYPES() -> List[int]:
```
make this a literal expression rather than a function

toloudis replied:
also the docstring says "get the number of" but it should say "get the list of".   I'm not sure why this is needed when you have the BINARY_BLOCK_TYPE enum as its own thing already.

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
         self._data = TrajectoryData.from_buffer_data(buffer_data)
 
+    @staticmethod
+    def _read_binary_file(input_file: InputFileData) -> Dict[str, Any]:
```
something about `_read_binary_file` vs `_load_binary`,  one of the two needs renaming because they sound the same.  Thinking out loud:  Maybe this one could be called `_get_data_views_from_file`.  or _read_file_into_memory_buffers

blairlyons replied:
renamed to `_binary_data_from_file()`

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+        result = {}
+        with open(input_file.file_path, "rb") as open_binary_file:
+            result["byte_view"] = open_binary_file.read()
+            result["n_values"] = int(
```
is this `n_values` actually used anywhere outside of here?  the file might contain encoded text data which does't really match up with BYTES_PER_VALUE.  Also why should this result be a dict rather than a namedTuple or some simple dataclass?

toloudis replied:
also wondering if struct.unpack copies the data?

toloudis replied:
see this as possible alternative to the use of struct.unpack in this case:  https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html

blairlyons replied:
I made a dataclass `BinaryFileData` and switched to using `numpy.frombuffer()` 🎉

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+        """
+        Parse header of data from a .simularium binary file
+        """
+        pos = len(BINARY_SETTINGS.HEADER)
```
a tiny bit more naming confusion: this function is called _binary_header and we are parsing this thing called BINARY_SETTINGS.HEADER and comparing it with a variable called binary_id.  I suggest the BINARY_SETTINGS.HEADER be renamed to BINARY_SETTINGS.FILE_IDENTIFIER or something like that.

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+        return result
+
+    @staticmethod
+    def _binary_header(binary_data: Dict[str, Any]) -> Dict[str, Any]:
```
this function could use a comment or docstring that describes the binary format of the header.  I really think you could get more readable code by just parsing the header using 2 or 3 struck.unpack calls on reads from the raw file bytes, for example in the way it's described in this post: https://stackoverflow.com/questions/30630090/parsing-binary-data-into-separate-variables-in-python

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+            "block_offsets": header_values[3::3],
+            "block_types": header_values[4::3],
+            "block_lengths": header_values[5::3],
+        }
```
again i feel this could be a dataclass or namedtuple rather than a dict

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+            raise DataError(
+                "Block type is not consistent for block "
+                f"#{block_index} : {block_type} != "
+                + binary_header["block_types"][block_index]
```
in these error messages it won't be clear which number came from the header and which came from the block

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+            )
+            + BINARY_SETTINGS.BLOCK_HEADER_N_VALUES
+        )
+        spatial_data_version = binary_data["int_view"][block_offset]
```
storing `binary_data["int_view"]` as some other variable name like `data_as_ints` (for example) would probably make this feel a bit more concise and readable.   because it shows up so many times

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+            "bundleStart": 0,
+            "bundleSize": n_frames,
+            "bundleData": [],
+        }
```
another dict that's really a dataclass or namedtuple

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
+                    block_index, binary_header, data
+                )
+            elif block_type == 3:
+                result["spatialData"] = FileConverter._binary_block_spatial_data(
```
what happens if there is a block of type 0 and a block of type 3 in the same file?  should we do a check in here and error out if that happens?    or things like two plotdata blocks or two trajectoryinfo blocks will always overwrite the result object.  worth noting in a comment at the very least.

toloudis commented on simulariumio/file_converter.py:
```diff
@@ -19,24 +21,207 @@
         """
-        This object loads the data in .simularium JSON format
-        at the input file
+        This object loads data from the input file in .simularium format.
```
Since you moved the writing code out into a `writers` section, why not move the reading code into a `readers` section?

toloudis commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,447 @@
+        """
+        Get the number of values in the bundle data buffer for each frame
+        """
+        total_steps = trajectory_data.agent_data.total_steps()
```
frames or steps?

toloudis commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,447 @@
+    def _header_n_bytes() -> int:
+        """
+        Get length of binary header in bytes
+        return 64 as long as there are 3 blocks, so doesn't need padding
```
is this comment about "return 64 as long as 3 blocks" even needed?

toloudis commented on simulariumio/readers/simularium_binary_reader.py:
```diff
@@ -0,0 +1,212 @@
+        Parse header of data from a .simularium binary file
+        """
+        id_length = len(BINARY_SETTINGS.FILE_IDENTIFIER)
+        pos = id_length + 3 * BINARY_SETTINGS.BYTES_PER_VALUE
```
Do you already have a symbolic constant for this `3` here?

toloudis commented on simulariumio/readers/simularium_binary_reader.py:
```diff
@@ -0,0 +1,212 @@
+                f"version > 1 up to current version {BINARY_SETTINGS.VERSION}, "
+                f"found {binary_version}"
+            )
+        block_info_length = 3 * BINARY_SETTINGS.BYTES_PER_VALUE * n_blocks
```
Do you already have a symbolic constant for this 3 here?

toloudis commented on simulariumio/readers/simularium_binary_reader.py:
```diff
@@ -0,0 +1,212 @@
+            block_type_id = SimulariumBinaryReader._binary_block_type(
+                block_index, block_info, binary_data.int_view
+            )
+            if block_type_id == 0:
```
isn't there a symbolic enum for these integer values?

toloudis commented on simulariumio/readers/simularium_binary_reader.py:
```diff
@@ -0,0 +1,212 @@
+                )
+            elif block_type == "spatialData":
+                result[
+                    "spatialData"
```
maybe change this to `block_type` to match line 195?

toloudis commented on simulariumio/writers/binary_chunk.py:
```diff
@@ -0,0 +1,56 @@
+        first_frame_index: int = 0,
+    ):
+        """
+        Object to store size info for a binary file
```
```suggestion
        Object to store size info for a binary file.  BinaryChunk is used to determine how many files to save and how to chunk up the data, if the data is larger than the max file size.
```

toloudis commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,441 @@
+        """
+        Get the number of values in the bundle data buffer for each frame
+        """
+        total_steps = trajectory_data.agent_data.total_timesteps()
```
If not changing this, then I think there needs to be some documentation somewhere that `frame` is synonymous with `timestep`, `step` and/or `time`.   When the code alternates between these terms interchangeably, I'd be worried that an engineer might be confused by lines like 
```
for frame_index in range(total_steps):
```
when it could easily be
```
for step_index in range(total_steps):
```

blairlyons replied:
I did change it `total_steps()` to `total_timesteps`. I could go through the codebase and remove `frame`, but I know it's used a bunch of places

toloudis replied:
can be separate PR/issue but I think it deserves consistency.  Also see this thread for how another team discussed what the data of a timestep is: https://github.com/MDAnalysis/mdanalysis/issues/250#issuecomment-98279984

toloudis commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,441 @@
+            spatial_header_n_values = (
+                BINARY_SETTINGS.BLOCK_HEADER_N_VALUES
+                + BINARY_SETTINGS.SPATIAL_BLOCK_HEADER_CONSTANT_N_VALUES
+                + 2 * (file_chunks[current_chunk].n_frames + 1)
```
is there already a symbolic constant for this `2`?

blairlyons replied:
no but I'll add one

toloudis commented on simulariumio/writers/binary_writer.py:
```diff
@@ -0,0 +1,441 @@
+                    block_n_bytes[1],
+                    block_offsets[2],
+                    block_types[2],
+                    block_n_bytes[2],
```
Just for fun, I googled "interleave multiple lists in python" and found this:
```
[val for tup in zip(block_offsets, block_types, block_n_bytes) for val in tup]
```
assuming the lists are all the same length... which would be future-proof for arbitrary numbers of blocks.

blairlyons replied:
cool!
